# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [85]:
import os
import numpy as np
import pandas as pd
from operator import itemgetter
import os
import glob

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [11]:
consecutive_integers = list(map(lambda x:x,range(1,51)))
print(consecutive_integers)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [13]:
even_numbers = list(map(lambda x:x*2,range(1,101)))
print(even_numbers)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [14]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [29]:
a.tolist()

[[0.84062117, 0.48006452, 0.7876326, 0.77109654],
 [0.44409793, 0.09014516, 0.81835917, 0.87645456],
 [0.7066597, 0.09610873, 0.41247947, 0.57433389],
 [0.29960807, 0.42315023, 0.34452557, 0.4751035],
 [0.17003563, 0.46843998, 0.92796258, 0.69814654],
 [0.41290051, 0.19561071, 0.16284783, 0.97016248],
 [0.71725408, 0.87702738, 0.31244595, 0.76615487],
 [0.20754036, 0.57871812, 0.07214068, 0.40356048],
 [0.12149553, 0.53222417, 0.9976855, 0.12536346],
 [0.80930099, 0.50962849, 0.94555126, 0.33364763]]

### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [28]:
a[a>=0.5]
    

array([0.84062117, 0.7876326 , 0.77109654, 0.81835917, 0.87645456,
       0.7066597 , 0.57433389, 0.92796258, 0.69814654, 0.97016248,
       0.71725408, 0.87702738, 0.76615487, 0.57871812, 0.53222417,
       0.9976855 , 0.80930099, 0.50962849, 0.94555126])

### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [53]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [58]:
b_list = b.tolist()
print(b_list)

[[[0.55867166, 0.06210792, 0.08147297], [0.82579068, 0.91512478, 0.06833034]], [[0.05440634, 0.65857693, 0.30296619], [0.06769833, 0.96031863, 0.51293743]], [[0.09143215, 0.71893382, 0.45850679], [0.58256464, 0.59005654, 0.56266457]], [[0.71600294, 0.87392666, 0.11434044], [0.8694668, 0.65669313, 0.10708681]], [[0.07529684, 0.46470767, 0.47984544], [0.65368638, 0.14901286, 0.23760688]]]


### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [79]:
def Extract(b): 
    return [item[1][-1] for item in b]

list_value = list(Extract(b))
last_value = []

for i in list_value:
    if i <= 0.5:
        last_value.append(i)
print(last_value)

[0.06833034, 0.10708681, 0.23760688]


### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

In [83]:
files = [f for f in os.listdir('/Users/alejandropalacios/Desktop/Ironhack/Data Analytics Bootcamp/Labs/Lab-list-comprehension/Lab-list-comprehension/lab-list-comprehensions/data') if f.endswith('.csv')]

print(files)

['sample_file_1.csv', 'sample_file_0.csv', 'sample_file_2.csv', 'sample_file_3.csv', 'sample_file_7.csv', 'sample_file_6.csv', 'sample_file_4.csv', 'sample_file_5.csv', 'sample_file_8.csv', 'sample_file_9.csv']


### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

In [93]:
path = r'/Users/alejandropalacios/Desktop/Ironhack/Data Analytics Bootcamp/Labs/Lab-list-comprehension/Lab-list-comprehension/lab-list-comprehensions/data'
all_files = glob.glob(path + "/*.csv")

li = []

for filename in all_files:
    df = pd.read_csv(filename, index_col = None, header = 0)
    li.append(df)

frame = pd.concat(li, axis = 0, ignore_index = True)

df.head(10)

0         1         2         3         4         5         6  \
0  0.215190  0.155352  0.160848  0.807736  0.363587  0.899832  0.146754   
1  0.895544  0.955196  0.089925  0.827555  0.089071  0.642883  0.996052   
2  0.413752  0.693052  0.789796  0.929164  0.536191  0.439769  0.773474   
3  0.728001  0.348156  0.935787  0.851163  0.444573  0.715080  0.988408   
4  0.134942  0.875931  0.273505  0.207588  0.080696  0.717396  0.033930   

          7         8         9        10        11        12        13  \
0  0.094802  0.705133  0.882762  0.773320  0.687745  0.016789  0.340725   
1  0.879020  0.421837  0.412141  0.858513  0.217091  0.176157  0.551236   
2  0.982074  0.876955  0.633154  0.279005  0.483317  0.908288  0.756172   
3  0.210332  0.732133  0.892383  0.216893  0.367595  0.846208  0.240111   
4  0.646837  0.888722  0.922742  0.176593  0.861333  0.389451  0.695244   

         14        15        16        17        18        19  
0  0.984182  0.985461  0.412044  0.867894  0.113432  0.349845  
1  0.834378  0.419535  0.041431  0.602258  0.984628  0.516899  
2  0.462130  0.289892  0.145233  0.076819  0.797836  0.197592  
3  0.471880  0.399721  0.758196  0.665568  0.931542  0.448124  
4  0.129955  0.364114  0.428224  0.365442  0.847818  0.588319

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.